src/extensions/default/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/PackageUpdaterStep.cs

```diff
@@ -150,6 +155,46 @@ void AddSubsteps<T>(IEnumerable<Operation<T>> items, string messageFormat, Func<
                 return new UpgradeStepInitializeResult(UpgradeStepStatus.Incomplete, $"{_analyzer.Name} has identified some recommended changes", Risk: _risk);
             }
 
+            // For server-only projects, do not need to add System.ServiceModel packages.
+            private static IEnumerable<Operation<NuGetReference>> UpdatePackageAddition(IDependencyCollection<NuGetReference> packages, IProject project)
+            {
+                var files = project.FindFiles(".cs", ProjectItemType.Compile);
+                var containsService = false;
+                if (files.Any())
+                {
+                    foreach (var f in files)
+                    {
+                        var root = CSharpSyntaxTree.ParseText(f).GetRoot();
+                        if (ContainsIdentifier(root, "ChannelFactory") || ContainsIdentifier(root, "ClientBase"))
+                        {
+                            return packages.Additions;
+                        }
+
+                        if (ContainsIdentifier(root, "ServiceHost"))
+                        {
+                            containsService = true;
+                        }
+                    }
+
+                    if (containsService)
+                    {
+                        return from p in packages.Additions
+                               where !p.Item.Name.Contains("System.ServiceModel")
+                               select p;
```

Comment:
Would StartsWith be suitable here? It's faster because it can skip most items immediately if they don't start with the letter S, whereas  Contains needs to search at least until there isn't enough space to contain the searched for string. Also should the overload be used which takes a StringComparison object. I don't know whether package names are case sensitive (I believe they aren't but not certain), in which case you should pass StringComparison.OrdinalIgnoreCase.